### PR TITLE
fix #1380, create NacosRegistryProviderObserver  when init method is executed

### LINKED ISSUE #1380: Nacos+SofaRpc 更新订阅端监听器 并发疑问
### Your question

- 场景：nacos+sofa的集成场景。

- 代码入口：com.alipay.sofa.rpc.registry.nacos.NacosRegistry#subscribe 

```java
                if (providerObserver == null) {
                    providerObserver = new NacosRegistryProviderObserver();
                }

                ProviderInfoListener providerInfoListener = config.getProviderInfoListener();
                providerObserver.addProviderListener(config, providerInfoListener);

                EventListener eventListener = new EventListener() {
                    @Override
                    public void onEvent(Event event) {
                        if (event instanceof NamingEvent) {
                            NamingEvent namingEvent = (NamingEvent) event;
                            List<Instance> instances = namingEvent.getInstances();
                            // avoid npe
                            if (null == instances) {
                                instances = new ArrayList<Instance>();
                            }
                            providerObserver.updateProviders(config, instances);
                        }
                    }
                };
                namingService.subscribe(serviceName, defaultCluster, eventListener);
                consumerListeners.put(config, eventListener);
```

上面的代码中，`providerObserver = new NacosRegistryProviderObserver();` 没有对`providerObserver`进行加锁控制，是否会出现并发的情况下（比如：多模块并行启动），存在多个服务订阅端，然后前面的一个Consumer创建了Observer类，然后后面的Consumer也创建了Observer类并进行覆盖，导致前面的Consumer listener丢失的情况。

### Environment

- SOFARPC version:
- JVM version (e.g. `java -version`):  1.8.0_332
- OS version (e.g. `uname -a`): Linux statement-group2-6c7c5df77d-vcqs4 5.4.247-162.350.amzn2.aarch64 #1 SMP Tue Jun 27 22:03:50 UTC 2023 aarch64 GNU/Linux
- Maven version: 3.8.4

### PR DESCRIPTION
**Pull request by wangchengming666**
fix #1380 修改之前的代码创建NacosRegistryProviderObserver是在subscribe里的，修改之后创建NacosRegistryProviderObserver是在init方法里的，init是有同步锁的。

### CODE REVIEW

File: registry/registry-nacos/src/main/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistry.java [hunk 1]
```diff
@@ -272,25 +286,28 @@ public List<ProviderGroup> subscribe(final ConsumerConfig config) {
             }
 
             try {
-                if (providerObserver == null) {
-                    providerObserver = new NacosRegistryProviderObserver();
+                ProviderInfoListener providerInfoListener = config.getProviderInfoListener();
+                try {
+                    lock.lock();
```
Comment (Lo1nt) [inline registry/registry-nacos/src/main/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistry.java:291]: better to leave `lock()` outside try block where `unlock()` appears.
ref to https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantLock.html

There are some discussion on this topic https://stackoverflow.com/questions/10868423/lock-lock-before-try

File: registry/registry-nacos/src/main/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistry.java [hunk 2]
```diff
@@ -147,6 +150,17 @@ public synchronized void init() {
         }
 
         try {
+            if (providerObserver == null) {
```
Comment (Lo1nt) [inline registry/registry-nacos/src/main/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistry.java:153]: Any test requires for these changes?

File: registry/registry-nacos/src/test/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistryTest.java [hunk 3]
```diff
@@ -77,7 +79,30 @@ public void setUp() {
     public void tearDown() {
         registry.destroy();
         registry = null;
-        serverConfig.destroy();
+    }
+
+    @Test
+    public void testMulitInit() throws InterruptedException {
```
Comment (EvenLjj) [inline registry/registry-nacos/src/test/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistryTest.java:85]: 拼接错误，testMuiltInit

File: registry/registry-nacos/src/main/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistry.java [hunk 4]
```diff
@@ -359,4 +354,14 @@ public void destroy() {
     public Properties getNacosConfig() {
         return nacosConfig;
     }
+
+    /**
+     * UT only
+     *
+     * @return
+     */
+    @Deprecated
```
Comment (EvenLjj) [inline registry/registry-nacos/src/main/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistry.java:363]: 可以使用 com.alipay.sofa.rpc.common.annotation.VisibleForTesting

File: registry/registry-nacos/src/test/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistryTest.java [hunk 5]
```diff
@@ -77,7 +79,30 @@ public void setUp() {
...
+        ExecutorService executorService = Executors.newFixedThreadPool(10);
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (EvenLjj) [inline registry/registry-nacos/src/test/java/com/alipay/sofa/rpc/registry/nacos/NacosRegistryTest.java:87]: 10个线程并行测试，这里要设置10

**Pull request closed without merging**